File: src/commands/bpb.ts
import { SlashCommandBuilder, ChatInputCommandInteraction, EmbedBuilder } from "discord.js";
import { reply } from "../lib/functions/discord";
import { parseTime } from "../lib/functions/util";
import { validateTarget } from "../lib/functions/schnose";
import { getMapKZGO, getMaps, getPB, getPlace, validateMap } from "gokz.js";
import userSchema from "../lib/schemas/user";
import modeMap from "gokz.js/lib/api";
import SchnoseBot from "src/classes/Schnose";

export default {
	data: new SlashCommandBuilder()
		.setName("bpb")
		.setDescription("Check a player's personal best on a bonus.")
		.addStringOption((o) => o.setName("map").setDescription("Specify a map.").setRequired(true))
		.addIntegerOption((o) => o.setName("course").setDescription("Specify a bonus."))
		.addStringOption((o) =>
			o.setName("mode").setDescription("Specify a mode.").setChoices(
				{
					name: "KZT",
					value: "kz_timer"
				},
				{
					name: "SKZ",
					value: "kz_simple"
				},
				{
					name: "VNL",
					value: "kz_vanilla"
				}
			)
		)
		.addStringOption((o) => o.setName("target").setDescription("Specify a target.")),

	async execute(interaction: ChatInputCommandInteraction, client: SchnoseBot) {
		await interaction.deferReply();

		const inputMap = interaction.options.getString("map")!;
		const inputCourse = interaction.options.getInteger("course") || 0;
		const inputMode = interaction.options.getString("mode") || null;
		const inputTarget = interaction.options.getString("target") || null;

		const globalMaps = await getMaps();
		if (!globalMaps.success) return reply(interaction, { content: globalMaps.error });

		const mapValidation = await validateMap(inputMap, globalMaps.data!);
		if (!mapValidation.success) return reply(interaction, { content: mapValidation.error });

		const KZGOMap = await getMapKZGO(mapValidation.data!.name);
		if (!KZGOMap.success) return reply(interaction, { content: KZGOMap.error });

		let mode: string;
		if (inputMode) mode = inputMode;
		else {
			const userDB = await userSchema.find({ discordID: interaction.user.id });
			if (!userDB[0]?.mode)
				return reply(interaction, {
					content: "You must either specify a mode or set a default value using `/mode`."
				});
			else mode = userDB[0].mode;
		}

		const targetValidation = await validateTarget(interaction, inputTarget);
		if (!targetValidation.success) return reply(interaction, { content: targetValidation.error });

		const req = await Promise.all([
			await getPB(targetValidation.data!.value!, mapValidation.data!.name, inputCourse, mode, true),
			await getPB(targetValidation.data!.value!, mapValidation.data!.name, inputCourse, mode, false)
		]);

		let tpPlace: any = null; // eslint-disable-line
		let proPlace: any | null = null; // eslint-disable-line
		if (req[0].success) tpPlace = await getPlace(req[0].data!);
		if (req[1].success) proPlace = await getPlace(req[1].data!);

		const embed = new EmbedBuilder()
			.setColor([116, 128, 194])
			.setTitle(
				`[BPB ${inputCourse}] ${
					req[0].data?.player_name || req[1].data?.player_name
						? `${req[0].data?.player_name || req[1].data?.player_name} on ${
								mapValidation.data!.name
						  }` // eslint-disable-line
						: `${mapValidation.data!.name}`
				}`
			)
			.setURL(`https://kzgo.eu/maps/${mapValidation.data!.name}`)
			.setThumbnail(
				`https://raw.githubusercontent.com/KZGlobalTeam/map-images/master/images/${
					mapValidation.data!.name
				}.jpg`
			)
			.addFields([
				{
					name: "TP",
					value: `${req[0].data?.time ? parseTime(req[0].data.time) : "😔"} ${
						tpPlace && tpPlace.success ? `(#${tpPlace?.data})` : `${req[0].success ? "?" : ""}`
					}`,
					inline: true
				},
				{
					name: "PRO",
					value: `${req[1].data?.time ? parseTime(req[1].data.time) : "😔"} ${
						proPlace && proPlace.success ? `(#${proPlace?.data})` : `${req[1].success ? "?" : ""}`
					}`,
					inline: true
				}
			])
			.setFooter({
				text: `(͡ ͡° ͜ つ ͡͡°)7 | ${modeMap.get(mode)}`,
				iconURL: client.icon
			});

		return reply(interaction, {
			embeds: [embed]
		});
	}
};

File: src/commands/bpb.rs
use {
	crate::{
		events::slash_commands::{InteractionState, InteractionResponseData::*},
		schnose::{InteractionResult, Target},
		db::retrieve_mode,
		util::*,
		gokz::{self, *},
	},
	futures::future::join_all,
	gokz_rs::{prelude::*, global_api::*},
	itertools::Itertools,
	serenity::{
		builder::{CreateApplicationCommand, CreateEmbed},
		model::prelude::command::CommandOptionType,
	},
};

pub(crate) fn register(cmd: &mut CreateApplicationCommand) -> &mut CreateApplicationCommand {
	return cmd
		.name("bpb")
		.description("Check a player's personal best on a bonus.")
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("map_name")
				.description("Specify a map.")
				.required(true)
		})
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("mode")
				.description("Choose a mode.")
				.add_string_choice("KZT", "kz_timer")
				.add_string_choice("SKZ", "kz_simple")
				.add_string_choice("VNL", "kz_vanilla")
				.required(false)
		})
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("player")
				.description("Specify a player.")
				.required(false)
		})
		.create_option(|opt| {
			opt.kind(CommandOptionType::Integer)
				.name("course")
				.description("Specify a course.")
				.required(false)
		});
}

pub(crate) async fn execute(state: &mut InteractionState<'_>) -> InteractionResult {
	// Defer current interaction since this could take a while
	state.defer().await?;

	let map_name = state.get::<String>("map_name").expect("This option is marked as `required`.");
	let course = state.get::<u8>("course").unwrap_or(1);
	let target = Target::from(state.get::<String>("player"));

	let player = target.to_player(state.user, state.db).await?;

	let mode = match state.get::<String>("mode") {
		Some(mode_name) => mode_name
			.parse()
			.expect("The possible values for this are hard-coded and should never be invalid."),
		None => retrieve_mode(state.user, state.db).await?,
	};

	let global_maps = get_maps(&state.req_client).await?;

	let map = is_global(&MapIdentifier::Name(map_name), &global_maps).await?;

	let map_identifier = MapIdentifier::Name(map.name.clone());

	let (tp, pro) = join_all([
		get_pb(&player, &map_identifier, &mode, true, course, &state.req_client),
		get_pb(&player, &map_identifier, &mode, false, course, &state.req_client),
	])
	.await
	.into_iter()
	.collect_tuple()
	.expect("This cannot fail, look 6 lines up.");

	if let (&Err(_), &Err(_)) = (&tp, &pro) {
		return Ok(Message("No BPBs found 😔.".into()));
	}

	let player_name = get_player_name((&tp, &pro));
	let (place_tp, place_pro) = (
		gokz::get_place(&tp, &state.req_client).await,
		gokz::get_place(&pro, &state.req_client).await,
	);
	let links = (gokz::get_replay_link(&tp).await, gokz::get_replay_link(&pro).await);

	let mut embed = CreateEmbed::default()
		.colour(state.colour)
		.title(format!(
			"[BPB {}] {} on {} (T{})",
			&course, &player_name, &map.name, &map.difficulty
		))
		.url(format!(
			"https://kzgo.eu/maps/{}?bonus={}&{}=",
			&map.name,
			&course,
			&mode.to_fancy().to_lowercase()
		))
		.thumbnail(state.thumbnail(&map.name))
		.field(
			"TP",
			format!(
				"{} {}",
				match &tp {
					Ok(rec) => format_time(rec.time),
					_ => String::from("😔"),
				},
				place_tp
			),
			true,
		)
		.field(
			"PRO",
			format!(
				"{} {}",
				match &pro {
					Ok(rec) => format_time(rec.time),
					_ => String::from("😔"),
				},
				place_pro
			),
			true,
		)
		.footer(|f| f.text(format!("Mode: {}", mode.to_fancy())).icon_url(&state.icon))
		.to_owned();

	attach_replay_links(&mut embed, links);

	return Ok(Embed(embed));
}
